Polkadot: Reverts return encoded error data (#1449)
hyperledger/solang: docs/targets/polkadot.rst

.. include:: ../examples/polkadot/call_flags.sol
  :code: solidity

Reverts and error data decoding
_______________________________

> When the contract decides to revert its execution, returned error data corresponds to what contracts on 

Reviewer: ```suggestion
When a contract reverts, the returned error data is what
```
(Removing superfluous words)

`EVM would return <https://docs.soliditylang.org/en/v0.8.20/control-structures.html#panic-via-assert-and-error-via-require>`_.
``assert()``, ``require()``, or ``revert()`` will revert the contract execution, where the revert reason 
is supplied as the contracts output (if any). Solidity contracts can also revert with a `Panic`.
Uncaught exceptions from calling and instantiating contracts or transferring funds will be bubbled 
up back to the caller.

The metadata contains all error variants that the contract knows about in the ``lang_error`` field.
The 4 bytes `selector` of the error data can be seen as the enum discriminator. However, because 
SCALE encoding does not allow discriminators larger than 1 byte, the hex-encoded error selector 
is provided as the enum variant name in the metadata (the selector could also be calculated by

> reconstructing and hashing the error signature based on the enum Variant types).

Author: What do you mean with built in? The error type is represented as an enum in the metadata, and it has variants.

> reconstructing and hashing the error signature based on the enum Variant types).

Reviewer 2: You wrote `Variant` with a capital V, as if its a name of something.

> reconstructing and hashing the error signature based on the enum Variant types).

Reviewer 2: That looks like `enum Variant`. If that is not what you mean, you should use a lower case letter.

> reconstructing and hashing the error signature based on the enum Variant types).

Author: Ah right, yeah that's true I'll fix it

> reconstructing and hashing the error signature based on the enum Variant types).

Reviewer: You could simply list the two error selectors for `Panic` and `Error`. 

Custom errors are not implemented in your PR.

> reconstructing and hashing the error signature based on the enum Variant types).

Author: Done that

.. include:: ../examples/polkadot/call_flags.sol
  :code: solidity

Reverts and error data decoding
_______________________________

When a contract reverts, the returned error data is what
`EVM would return <https://docs.soliditylang.org/en/v0.8.20/control-structures.html#panic-via-assert-and-error-via-require>`_.
``assert()``, ``require()``, or ``revert()`` will revert the contract execution, where the revert reason 

> is supplied as the contracts output (if any). Solidity contracts can also revert with a `Panic`.

Reviewer: Some words about where `Panic` occurs would be helpful

Author: Gonna add a link to the Solidity docs

is provided as the enum variant name in the metadata (the selector could also be calculated by
reconstructing and hashing the error signature based on the enum variant types).

.. note::

    Solidity knows about `Error`, `Panic` and 
    `custom errors <https://docs.soliditylang.org/en/latest/abi-spec.html#errors>`_.
    Solang does not yet support custom errors. For now, only `Error` (selector of `0x08c379a0`) 
    and `Panic` (selector of `4e487b71`) will be returned and occur in the metadata.

The general process of decoding the output data for Solang Solidity contracts is as follows:

1. If the revert flag is **not** set, the contract didn't revert and the output should be encoded as specified in the message spec.
2. The compiler version must be solang > 0.3.1, or the error data can't be decoded (check the ``compiler`` field in the contract metadata).
3. If the output length is smaller than 4 bytes, the error data can't be decoded (contracts may return empty error data, for example if ``revert()`` without arguments is used).
4. If the first 4 bytes of the output do **not** match any of the selectors found in ``lang_error``, the error can't be decoded.

> 5. **Skip** the selector (first 4 bytes) and decode the remaining data according to the corresponding variant type found in `lang_error`.

Reviewer: Some example code would be great

Author: What code example do you have in mind?

I don't see what a code example would add. Also, a Typescript or Rust snippet to do this will be quite long I don't see why this belongs into our docs, a solidity example is uselesse because try-catch should be used anyways.

Reviewer: How about a link to an integration test?

Author: Well, in the integration test we always just match the output directly against the expected output bytes (integration tests are not examples...)

.. include:: ../examples/polkadot/call_flags.sol
  :code: solidity

Reverts and error data decoding
_______________________________

When a contract reverts, the returned error data is what
`EVM would return <https://docs.soliditylang.org/en/v0.8.20/control-structures.html#panic-via-assert-and-error-via-require>`_.
``assert()``, ``require()``, or ``revert()`` will revert the contract execution, where the revert reason 
is supplied as the contracts output (if any). Solidity contracts can also revert with a `Panic` 
(please refer to the 

> `Solidity language documentation <https://docs.soliditylang.org/en/v0.8.20/control-structures.html#panic-via-assert-and-error-via-require>`_.

Reviewer: Please call this `Ethereum Solidity documentation. It's totally confusing to call this "Solidity language documentation".

Author: Why is it confusing calling the Solidity Language Documentation the Solidity Language Documentation.

If anything, `Ethereum Solidity documentation` is confusing. What is `Ethereum Solidity`? What is not `Ethereum Solidity`?

I don't get why we can't call the Solidity Language by it's name.

You go to a website, the website is called "https://soliditylang.org/".

This implies that there is a thing called "Solidity Language".

On this website, there is a button, called "read the docs", leading to another website at "https://docs.soliditylang.org".

So, if there is the `Solidity Langauge`, and it has a `Documentation`. How is it "totally confusing" to calling it "Solidity language documentation"?

for more information about when `Panic` might be returned).
Uncaught exceptions from calling and instantiating contracts or transferring funds will be bubbled 
up back to the caller.

The metadata contains all error variants that the contract knows about in the ``lang_error`` field.
The 4 bytes `selector` of the error data can be seen as the enum discriminator. However, because 
SCALE encoding does not allow discriminators larger than 1 byte, the hex-encoded error selector 
is provided as the enum variant name in the metadata (the selector could also be calculated by

> reconstructing and hashing the error signature based on the enum variant types).

Reviewer: So where does the name go? E.g. `Panic`, `Error`, or `CustomError5`? 

I do think this this paragraph is inscrutable. I still don't have a good idea of what is going on here.

Author: Makes me think, do you think it'd be better if the Selector goes into the path of the Variant Type?

.. note::

    Solidity knows about `Error`, `Panic` and 

>     `custom errors <https://docs.soliditylang.org/en/latest/abi-spec.html#errors>`_.

Reviewer: Please don't refer to solc/Ethereum Solidity as "Solidity". That suggests we are not Solidity

Author: > Please don't refer to solc/Ethereum Solidity as "Solidity"

I disagree. Solidity is what the language authors think it is. And that happens to be whats on soliditylang.org. We can't say we are a Solidity compiler when we in fact compile something that is not Solidity.

If you mean `solc/Ethereum` what is broadly understood as the _Solidity Language_, it doesn't follow to not call it _Solidity_. Saying  "Solidity knows about `X`" is correct.

Reviewer: Also, I've called it "Ethereum Solidity" throughout the documentation. If you're going to change this, then change all the links that say "Ethereum Solidity".

Reviewer 2: In one hand, Solidity for Solana has its own syntax, so we can see it as a different flavor. Calling Ethereum Solidity to refer to the original Solidity implementation is reasonable, given that we are slightly different than the original.

Reviewer: > In one hand, Solidity for Solana has its own syntax, so we can see it as a different flavor. Calling Ethereum Solidity to refer to the original Solidity implementation is reasonable, given that we are slightly different than the original.

Exactly. On Polkadot there are also many small differences with Ethereum Solidity (or should be called Solidity on Ethereum/ Solidity on EVM)?

Author: > Your link suggests that https://docs.soliditylang.org/en/latest/abi-spec.html#errors is the authoritative answer.

Yes, intentionally so. I'd be largely surprised if this wasn't the authoritative answer. For a vast majority of anyone referring to the Solidity Language in any way.

> Where does that leave Solang?

This is concisely explained in literally the next sentence:

    Solang does not yet support custom errors.

> Also, I've called it "Ethereum Solidity" throughout the documentation. 

Not true. By just having a very quick look at the docs, I find _many_ instances we just talk about `Solidity`. No mention of `Ethereum Solidity`, `Solang Solidity` or whathever else. Just `Solidity`. And in this case I assume anyone reaading the docs will `Solidity` as to what's documented on docs.soliditylang.org and implemented by solc.

For example in the Solana section. The first sentence:

    As the underlying Solana environment is different than that of Ethereum, Solidity inner workings have been modified to function properly.

So, this should mention that the inner workings of `Ethereum Solidity` have been modified, not `Solidity`. There are _many_ more of these instances.

> If you're going to change this, then change all the links that say "Ethereum Solidity"

As pointed out, in our docs, this is not at all consistent. In fact, there are many instances we write just `Solidity`, without referring to either what you mean with "Ethereum Solidity", nor to "Solang Solidity" nor "Solidity on Solana" nor anything else. So why not do the only sane thing here which is referring the term `Solidity` as to what everyone else in the world refers it to.

Author: > > In one hand, Solidity for Solana has its own syntax, so we can see it as a different flavor. Calling Ethereum Solidity to refer to the original Solidity implementation is reasonable, given that we are slightly different than the original.
> 
> Exactly. On Polkadot there are also many small differences with Ethereum Solidity (or should be called Solidity on Ethereum/ Solidity on EVM)?

I'm fine with calling it `Solana Solidity` or `Solidity on Solana` or `Solang Solidity` or whatever. I'm just not convinced that we can not just use `Solidity` when referring to what the World understands as `Solidity` to be is what's documented on soliditylang.org and in `solc`.

And this is _very_ inconsistent in our docs.

But fine. Apparently it's the end of the world if someone talks about `Solidity` in our docs. I'm just gonna `s/Solidity/Ethereum Solidity` here to end this stupid discussion for good

Author: Ah, now I get what you mean.

But it doesn't work like that.

You can't invent random German sounding words and start using in them in a local community still call your Language German.

What you end up is a _dialect_.

So the correct thing to say would be that Solang is a Compiler for _Solidity Dialects_ running on other runtimes.

But that still leaves the solc/docs.soliditylang.org the authority. We can't just deny that fact.

Reviewer: I am fine with dialect. 

> But that still leaves the solc/docs.soliditylang.org the authority. We can't just deny that fact.

I am totally denying that fact. 

Just in the same way. there is no version German which is not a dialect. Each and every one is a dialect. Just in the same way, the King of England speaks a dialect.

Author: I summarize this discussion into the following points:

- Neither the `solc` compiler nor the `docs.soliditylang.org` resemble a language spec.
- Hence it's questionable whether this is authoritative because it's merely the predominant dialect or version of Solidity at the time of writing.
- Add to that, seeing the compiler or docs of a version of Solidity that can only run on a specific runtime is definitively not what we want.
- From that follows, saying just `Solidity` while linking to some specific documentation or implementation should be avoided. Because it implies that the documentation or implementation linked to to be an authoritative source while it is not.

Author: @seanyoung we can (we should!) continue discussion in #1479

as for this PR I can just change it to `Ethereum Solidity` because that's what seems to be used most often in other places.

    Solang does not yet support custom errors. For now, only `Error` (selector of `0x08c379a0`) 

>     and `Panic` (selector of `4e487b71`) will be returned and occur in the metadata.

Reviewer: Add `0x` prefix to 4e487b71 to make it consistent with other selector
